# List Comprehensions Lab

Complete the following set of exercises to solidify your knowledge of list comprehensions.

In [1]:
import os
import numpy as np
import pandas as pd

### 1. Use a list comprehension to create and print a list of consecutive integers starting with 1 and ending with 50.

In [31]:
print([e for e in range(1,51)])

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50]


### 2. Use a list comprehension to create and print a list of even numbers starting with 2 and ending with 200.

In [34]:
print([e for e in range(2,201) if e%2 == 0])

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 28, 30, 32, 34, 36, 38, 40, 42, 44, 46, 48, 50, 52, 54, 56, 58, 60, 62, 64, 66, 68, 70, 72, 74, 76, 78, 80, 82, 84, 86, 88, 90, 92, 94, 96, 98, 100, 102, 104, 106, 108, 110, 112, 114, 116, 118, 120, 122, 124, 126, 128, 130, 132, 134, 136, 138, 140, 142, 144, 146, 148, 150, 152, 154, 156, 158, 160, 162, 164, 166, 168, 170, 172, 174, 176, 178, 180, 182, 184, 186, 188, 190, 192, 194, 196, 198, 200]


### 3. Use a list comprehension to create and print a list containing all elements of the 10 x 4 Numpy array below.

In [46]:
a = np.array ([[0.84062117, 0.48006452, 0.7876326 , 0.77109654],
               [0.44409793, 0.09014516, 0.81835917, 0.87645456],
               [0.7066597 , 0.09610873, 0.41247947, 0.57433389],
               [0.29960807, 0.42315023, 0.34452557, 0.4751035 ],
               [0.17003563, 0.46843998, 0.92796258, 0.69814654],
               [0.41290051, 0.19561071, 0.16284783, 0.97016248],
               [0.71725408, 0.87702738, 0.31244595, 0.76615487],
               [0.20754036, 0.57871812, 0.07214068, 0.40356048],
               [0.12149553, 0.53222417, 0.9976855 , 0.12536346],
               [0.80930099, 0.50962849, 0.94555126, 0.33364763]])

In [118]:
print([a[i,j] for i in range(a.shape[0]) for j in range(a.shape[1])])

[0.84062117, 0.48006452, 0.7876326, 0.77109654, 0.44409793, 0.09014516, 0.81835917, 0.87645456, 0.7066597, 0.09610873, 0.41247947, 0.57433389, 0.29960807, 0.42315023, 0.34452557, 0.4751035, 0.17003563, 0.46843998, 0.92796258, 0.69814654, 0.41290051, 0.19561071, 0.16284783, 0.97016248, 0.71725408, 0.87702738, 0.31244595, 0.76615487, 0.20754036, 0.57871812, 0.07214068, 0.40356048, 0.12149553, 0.53222417, 0.9976855, 0.12536346, 0.80930099, 0.50962849, 0.94555126, 0.33364763]


### 4. Add a condition to the list comprehension above so that only values greater than or equal to 0.5 are printed.

In [120]:
print([a[i,j] for i in range(a.shape[0]) for j in range(a.shape[1]) if a[i,j] >= 0.5])

[0.84062117, 0.7876326, 0.77109654, 0.81835917, 0.87645456, 0.7066597, 0.57433389, 0.92796258, 0.69814654, 0.97016248, 0.71725408, 0.87702738, 0.76615487, 0.57871812, 0.53222417, 0.9976855, 0.80930099, 0.50962849, 0.94555126]


### 5. Use a list comprehension to create and print a list containing all elements of the 5 x 2 x 3 Numpy array below.

In [238]:
b = np.array([[[0.55867166, 0.06210792, 0.08147297],
        [0.82579068, 0.91512478, 0.06833034]],

       [[0.05440634, 0.65857693, 0.30296619],
        [0.06769833, 0.96031863, 0.51293743]],

       [[0.09143215, 0.71893382, 0.45850679],
        [0.58256464, 0.59005654, 0.56266457]],

       [[0.71600294, 0.87392666, 0.11434044],
        [0.8694668 , 0.65669313, 0.10708681]],

       [[0.07529684, 0.46470767, 0.47984544],
        [0.65368638, 0.14901286, 0.23760688]]])

In [239]:
print([b[i,j,k] for i in range(b.shape[0]) for j in range(b.shape[1]) for k in range(b.shape[2])])

[0.55867166, 0.06210792, 0.08147297, 0.82579068, 0.91512478, 0.06833034, 0.05440634, 0.65857693, 0.30296619, 0.06769833, 0.96031863, 0.51293743, 0.09143215, 0.71893382, 0.45850679, 0.58256464, 0.59005654, 0.56266457, 0.71600294, 0.87392666, 0.11434044, 0.8694668, 0.65669313, 0.10708681, 0.07529684, 0.46470767, 0.47984544, 0.65368638, 0.14901286, 0.23760688]


### 5. Add a condition to the list comprehension above so that the last value in each subarray is printed, but only if it is less than or equal to 0.5.

In [241]:
print([y[-1] for x in b for y in x if y[-1] <= 0.5])

[0.08147297, 0.06833034, 0.30296619, 0.45850679, 0.11434044, 0.10708681, 0.47984544, 0.23760688]


### 6. Use a list comprehension to select and print the names of all CSV files in the */data* directory.

In [255]:
csv_list = ([f for f in os.listdir('../data') if f.endswith('.csv')])

['sample_file_6.csv',
 'sample_file_2.csv',
 'sample_file_3.csv',
 'sample_file_1.csv',
 'sample_file_5.csv',
 'sample_file_7.csv',
 'sample_file_4.csv',
 'sample_file_9.csv',
 'sample_file_0.csv',
 'sample_file_8.csv']

### 7. Use a list comprehension and the Pandas `read_csv` and `concat` methods to read all CSV files in the */data* directory and combine them into a single data frame. Display the top 10 rows of the resulting data frame.

In [270]:
import os
import pandas as pd

csv_list = [f for f in os.listdir('../data') if f.endswith('.csv')]
data_sets = [pd.read_csv(os.path.join('../data', f)) for f in csv_list]
data = pd.concat(data_sets, axis=0)

data.iloc[:10]

0         1         2         3         4         5         6  \
0  0.702566  0.092336  0.245083  0.918402  0.695110  0.886921  0.588605   
1  0.177005  0.649551  0.467321  0.449597  0.458872  0.815586  0.455132   
2  0.689556  0.445586  0.005691  0.150171  0.554146  0.492320  0.185612   
3  0.047188  0.854596  0.833368  0.043878  0.883505  0.786302  0.980916   
4  0.936065  0.725831  0.184239  0.677391  0.242726  0.125457  0.356254   
0  0.948664  0.215285  0.918270  0.599951  0.755120  0.971609  0.103190   
1  0.163236  0.803926  0.916655  0.775234  0.644890  0.701362  0.910208   
2  0.934136  0.031410  0.954057  0.853387  0.642160  0.681184  0.317198   
3  0.757038  0.918964  0.475459  0.837686  0.149645  0.819032  0.611996   
4  0.263455  0.816283  0.336707  0.587997  0.285871  0.619942  0.018027   

          7         8         9        10        11        12        13  \
0  0.950970  0.168079  0.783107  0.698836  0.120857  0.207432  0.824789   
1  0.729812  0.998528  0.048724  0.377391  0.729498  0.721515  0.566724   
2  0.060855  0.457443  0.847337  0.603072  0.346574  0.898464  0.482963   
3  0.004030  0.619601  0.547769  0.160962  0.545281  0.931623  0.991855   
4  0.937350  0.215774  0.212762  0.483018  0.573162  0.634487  0.423924   
0  0.194754  0.932388  0.591727  0.697517  0.607355  0.177649  0.435968   
1  0.871204  0.321745  0.586035  0.887054  0.240060  0.915342  0.205310   
2  0.875259  0.538416  0.867511  0.813309  0.215624  0.552062  0.498378   
3  0.644348  0.938444  0.410444  0.561513  0.499231  0.856437  0.054619   
4  0.548845  0.121471  0.194299  0.149844  0.848866  0.531840  0.663384   

         14        15        16        17        18        19  
0  0.349212  0.818606  0.376941  0.886644  0.472826  0.551858  
1  0.338366  0.877239  0.129824  0.773155  0.485385  0.686294  
2  0.443979  0.430132  0.713078  0.302786  0.051382  0.203699  
3  0.821099  0.654592  0.346710  0.183948  0.372739  0.739383  
4  0.394633  0.766967  0.606719  0.538712  0.885516  0.667051  
0  0.202404  0.979777  0.095713  0.159040  0.651457  0.803393  
1  0.489504  0.848926  0.304342  0.358977  0.841539  0.964889  
2  0.739656  0.307914  0.233996  0.602166  0.244210  0.313071  
3  0.326310  0.461825  0.954783  0.361873  0.145952  0.873029  
4  0.084884  0.120312  0.463214  0.437889  0.542376  0.668447

### 8. Use a list comprehension to select and print the column numbers for columns from the data set whose median is less than 0.48.

### 9. Use a list comprehension to add a new column (20) to the data frame whose values are the values in column 19 minus 0.1. Display the top 10 rows of the resulting data frame.

### 10. Use a list comprehension to extract and print all values from the data set that are between 0.7 and 0.75.